### PR TITLE
[WIP] Puppeteer backend

### PR DESCRIPTION
**Pull request by barryvdh**
This isn't much tested yet, more a proof-of-concept.

Uses https://github.com/GoogleChrome/puppeteer (see #284 )

Installation:
Requires Node 6.4+ or 7+

```
sudo apt-get install -y nodejs gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget
sudo npm install --global --unsafe-perm puppeteer
sudo chmod -R o+rx /usr/lib/node_modules/puppeteer/.local-chromium
```

(Chrome is installed by Puppeteer by default, but needs to be executable. There also is an option to use your own Chrome, but could be out sync)

Example

```php
require __DIR__ . '/../vendor/autoload.php';

$snappy = new \Knp\Snappy\Puppeteer\Pdf();

$result = $snappy->getOutput('http://www.github.com', ['format' => 'a6']);
header('Content-Type: application/pdf');
header('Content-Disposition: inline; filename="file.pdf"');
echo $result;

$snappy->generate('http://www.github.com', 'github.pdf', [], true);

$snappy->generateFromHtml('<h1>Bill</h1><p>You owe me money, dude.</p>', '/test.pdf');
```

Not all options are configured/tested yet, but should provide some more options then the CLI.

Only thing I saw is that all HTML is now passed through the commandline? Not sure if that has limitations. Also the chrome generator added `file://` in front of the file, which doesn't work for urls.

The script is based on https://github.com/spatie/browsershot/tree/v3 but extended for more options and modified for Node6

### CODE REVIEW

**Comment by barryvdh** [comment]
To be fair, this does seem to be slower . Not tested with large ones, but wkhtmltopdf + chrome cli are 300 ms for just the `'<h1>Bill</h1><p>You owe me money, dude.</p>'` html, and puppeteer is around 600. But I suspect this is mainly initial overhead for the node script. And this doesn't have the title above the pdf for example. So flexibility over performance for this.

**Comment by barryvdh** [comment]
Simplified it a bit, should also work without installing it globally, by passing the directory in which it is installed (eg. the root of the project).

```php
$image = new \Knp\Snappy\Puppeteer\Image(['type' => 'png'],  __DIR__ .'/..');
```

**Comment by barryvdh** [comment]
Add validation for the currently supported types. 

So is this something your interested in? Otherwise I'll create a standalone package or something.

File: src/Knp/Snappy/Puppeteer/AbstractGenerator.php [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        if ($this->nodePath && is_dir($this->nodePath)) {
+            $nodePath = escapeshellarg($this->nodePath);
+        } else {
+            $nodePath = '`npm root -g`';    // Detect root node path
+        }
+
+        return implode(' ', [
+            'NODE_PATH=' . $nodePath,
+            'node',
+            escapeshellarg(dirname(dirname(dirname(dirname(__DIR__)))) . '//resources/puppeteer.js'),
```
Comment (NiR-) [inline src/Knp/Snappy/Puppeteer/AbstractGenerator.php]: Why don't you use the more common `../../../../`? Due to windows?

**Comment by barryvdh** [comment]
No special reason, just a shorter command path

File: src/Knp/Snappy/Puppeteer/AbstractGenerator.php (hunk 1 above)
Comment (NiR-) [reply src/Knp/Snappy/Puppeteer/AbstractGenerator.php]: Could you do only one call to `dirname` instead please?

**Comment by NiR-** [comment]
Could you fix my last comments and the little CS issues please? After that, I'd be happy to merge this PR :)

**Comment by Bavragor** [comment]
Maybe this could help https://github.com/nesk/puphpeteer

**Pull request closed without merging**